File: nodejs/exchange.js
import { EventEmitter } from 'events';

// Base Exchange interface.
// To swap for a real exchange (Binance, Alpaca, Coinbase, etc.):
//   1. Subclass `Exchange` (e.g. `class BinanceExchange extends Exchange`).
//   2. In start(), open a WebSocket to the venue's market-data feed.
//   3. On each incoming message, emit a 'tick' event: { symbol, price, ts }.
//   4. In submitOrder(), POST to the REST order endpoint and return the fill.
export class Exchange extends EventEmitter {
  async submitOrder(order) { throw new Error('submitOrder: not implemented'); }
  start(intervalMs) { throw new Error('start: not implemented'); }
  stop() { throw new Error('stop: not implemented'); }
}

// MockExchange: simulates a market feed with a random walk and fills orders
// instantly at the latest price (zero slippage, zero latency).
export class MockExchange extends Exchange {
  constructor(symbol, basePrice) {
    super();
    this.symbol = symbol;
    this.price = basePrice;
    this.timer = null;
  }

  start(intervalMs = 100) {
    this.timer = setInterval(() => {
      // Random walk: price moves +/- 0.1% per tick.
      const drift = (Math.random() - 0.5) * 0.002 * this.price;
      this.price += drift;
      this.emit('tick', { symbol: this.symbol, price: this.price, ts: Date.now() });
    }, intervalMs);
  }

  stop() {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
  }

  async submitOrder(order) {
    // Mock fill at current price. Real exchange would return fill price from API.
    return { ...order, fillPrice: this.price, status: 'filled', ts: Date.now() };
  }
}

File: python/exchange.py
"""Exchange abstraction + mock implementation."""
import asyncio
import random
import time
from dataclasses import dataclass
from typing import AsyncIterator


@dataclass
class Tick:
    symbol: str
    price: float
    ts: float


@dataclass
class Order:
    symbol: str
    side: str  # 'BUY' or 'SELL'
    qty: float


@dataclass
class Fill:
    order: Order
    fill_price: float
    ts: float


class Exchange:
    """Base interface.

    To swap in a real exchange (Binance, Alpaca, Coinbase, ...):
      - Implement `ticks()` as an async generator that reads a WebSocket
        feed (e.g. using the `websockets` library) and yields `Tick`s.
      - Implement `submit_order()` to POST to the venue's REST API
        (e.g. via `aiohttp`) and return a `Fill`.
    """

    async def ticks(self) -> AsyncIterator[Tick]:
        raise NotImplementedError
        yield  # pragma: no cover  (makes this an async generator)

    async def submit_order(self, order: Order) -> Fill:
        raise NotImplementedError


class MockExchange(Exchange):
    """Simulates a market feed via random walk, fills orders instantly."""

    def __init__(self, symbol: str, base_price: float, interval_ms: int = 100):
        self.symbol = symbol
        self.price = base_price
        self.interval = interval_ms / 1000.0
        self._running = True

    def stop(self) -> None:
        self._running = False

    async def ticks(self) -> AsyncIterator[Tick]:
        while self._running:
            # Random walk: price moves +/- 0.1% per tick.
            self.price += (random.random() - 0.5) * 0.002 * self.price
            yield Tick(self.symbol, self.price, time.time())
            await asyncio.sleep(self.interval)

    async def submit_order(self, order: Order) -> Fill:
        # Mock fill at current price. Zero slippage, zero latency.
        return Fill(order, self.price, time.time())
